### DISCUSSION TITLE
Getting segmentation fault while sorting parquet file with more than 1 billion rows

### DISCUSSION THREAD

**Discussion by mayankanand007** [duckdb/duckdb#9452] [General]
Hey team! Thanks for all the work on `DuckDB`! 

I'm trying to add `order by` to my `duckdb` query which is expected to produce more than one billion rows in its final parquet file-based output. I'm trying to order by multiple columns (3), but when I add the `order by` I end up getting a segmentation fault. The fix so far has been to reduce the number of rows I have to sort by running my query in batches. 

The issue is that so far, I have only reproduced this issue in an internal organizational dev environment, while I'm unable to do that on my personal laptop. When I try to sort 1 billion rows on my M1 Mac (16GB), it just throws OOM issue, while in the internal environment running on [OCP](https://www.redhat.com/en/technologies/cloud-computing/openshift/container-platform) I see segmentation fault.

Due to restrictions, I can't share the data or the code/query, but I'm hoping I can get a better understanding as to some reasons this issue can happen, or some directions I can explore, as I'm not super familiar with `duckdb` internals.

Thank you for your time!

**Comment by l1t1** [comment]
add memery to your computer

**Reply by Mause** [reply]
This is not a helpful suggestion for anyone with a modern laptop, as they're not upgradable.

**Reply by mayankanand007** [reply]
Running on OCP (where I see the segfault) - I have up to 256GB available and 2 CPUs, but I do not see any spike in either of their usages. I believe if memory is an issue, it should throw OOM?

The usage stays around 60 - 70G and never goes above 100G from what I remember. Not sure if there is something obvious I'm missing?

**Comment by Mause** [comment]
Are you using an in memory connection with a temporary directory set, or an on disk database?

**Reply by mayankanand007** [reply]
I'm currently using an in-memory `:memory:` connection and have set a temporary directory as well with up to 100G of space using the openshift PVC (https://docs.openshift.com/container-platform/4.8/storage/understanding-persistent-storage.html)

**Reply by mayankanand007** [reply]
The parquet file that I'm trying to sort is ~ 10G and essentially has a structure similar to client transactions with a client ID, amount, date, merchant code, etc. I was trying to sort by client Id, date and amount

**Comment by l1t1** [comment]
could you post the result of `select * from 'unsorted parquet file' where 1=0`? it's useful to reproduce the issue.

**Comment by mayankanand007** [comment]
Thanks for the advice here! I tried to re-run my code but instead of an in-memory connection with a temporary directory, I'm now using an on-disk connection. This seems to solve my issue and I no longer get a segmentation fault while sorting 🎉 

Is it because an on-disk connection allows it to run external sort more easily?

**Reply by Mause** [reply]
Generally speaking yes, it allows DuckDB to offload data it isn't using to disk which frees up ram for other tasks


For my own curiosity, what's the exact segfault message you're seeing?

**Reply by mayankanand007** [reply]
Thank you for your response, here is the error message:

```bash
[2023-10-13, 16:50:15 UTC] {pod_manager.py:228} INFO - /tmp/tmpjgmghbjg: line 3:    17 Segmentation fault      python -m pex_scripts.consolidate_filter
```

**Reply by mayankanand007** [reply]
btw, could you also explain to me the difference between using an in-memory connection with a temporary directory set, or an on-disk database?

what is the use of setting the temp directory?
